File: app/src/main/java/com/restart/spacestationtracker/adapter/PeopleInSpaceAdapter.java
package com.restart.spacestationtracker.adapter;

import android.annotation.SuppressLint;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.browser.customtabs.CustomTabColorSchemeParams;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.restart.spacestationtracker.R;
import com.restart.spacestationtracker.data.Astronaut;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * The type People in space adapter.
 */
public class PeopleInSpaceAdapter extends RecyclerView.Adapter<PeopleInSpaceAdapter.PeopleInSpaceAdapterViewHolder> {

    private final AppCompatActivity mActivity;
    private final SimpleDateFormat mDateFormat;
    private List<Astronaut> mAstronauts;

    /**
     * Instantiates a new People in space adapter.
     *
     * @param activity   The calling activity
     * @param astronauts The incoming list of astronauts to be displayed
     */
    public PeopleInSpaceAdapter(AppCompatActivity activity, List<Astronaut> astronauts) {
        mActivity = activity;
        mAstronauts = astronauts;
        mDateFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault());
    }

    /**
     * The type People in space adapter view holder.
     */
    static class PeopleInSpaceAdapterViewHolder extends RecyclerView.ViewHolder {
        /**
         * All widgets that will hold values unique to an astronaut
         */
        private final ImageView mProfileImage;
        private final ProgressBar mAstronautPictureProgress;
        private final ImageView mCountryFlag;
        private final ImageView mAstronautTwitter;
        private final ImageView mAstronautFacebook;
        private final ImageView mAstronautInstagram;
        private final ImageView mAstronautWiki;
        private final ImageView mAstronautGoogle;
        private final TextView mName;
        private final TextView mRole;
        private final TextView mDate;
        private final TextView mBio;

        /**
         * Instantiates a new People in space adapter view holder.
         *
         * @param view The incoming parent view
         */
        PeopleInSpaceAdapterViewHolder(View view) {
            super(view);

            /* Get references to each view. */
            mName = view.findViewById(R.id.name);
            mRole = view.findViewById(R.id.role);
            mDate = view.findViewById(R.id.days_since_launch);
            mBio = view.findViewById(R.id.bio);
            mProfileImage = view.findViewById(R.id.img);
            mCountryFlag = view.findViewById(R.id.countryFlag);
            mAstronautTwitter = view.findViewById(R.id.astronautTwitter);
            mAstronautInstagram = view.findViewById(R.id.astronautInstagram);
            mAstronautFacebook = view.findViewById(R.id.astronautFacebook);
            mAstronautWiki = view.findViewById(R.id.astronautWiki);
            mAstronautPictureProgress = view.findViewById(R.id.astronaut_picture_progress);
            mAstronautGoogle = view.findViewById(R.id.astronautGoogle);
        }
    }

    /**
     * Inflate the new list view.
     *
     * @param parent   To inflate the view
     * @param viewType N/A
     * @return New inflated view
     */
    @NonNull
    @Override
    public PeopleInSpaceAdapter.PeopleInSpaceAdapterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new PeopleInSpaceAdapterViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.people_row, parent, false));
    }

    /**
     * Setup the values for each widgets. Use the incoming ViewHolder to gain access to their corresponding
     * references.
     *
     * @param holder   ViewHolder containing the widget references
     * @param position The index in our list being updated with new data
     */
    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull final PeopleInSpaceAdapterViewHolder holder, int position) {
        final Astronaut astronaut = mAstronauts.get(position);

        /* If an astronaut exists in the data set, let update the values */
        if (astronaut != null) {
            String location;
            if (astronaut.getIsIss()) location = "ISS";
            else location = "Tiangong-2";

            // Does the astronaut have a twitter handle?
            if (!astronaut.getTwitter().isEmpty()) {
                holder.mAstronautTwitter.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_twitter));
            } else {
                holder.mAstronautTwitter.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_twitter_grey));
            }

            holder.mAstronautTwitter.setOnClickListener(view -> {
                if (!astronaut.getTwitter().isEmpty()) {
                    startCustomTab(astronaut.getTwitter());
                } else {
                    Toast.makeText(mActivity, astronaut.getName().split(" ")[0] + " " + mActivity.getString(R.string.errorNoTwitter), Toast.LENGTH_SHORT).show();
                }
            });

            // Does the astronaut have a instagram handle?
            if (!astronaut.getInstagram().isEmpty()) {
                holder.mAstronautInstagram.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_instagram));
            } else {
                holder.mAstronautInstagram.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_instagram_grey));
            }

            holder.mAstronautInstagram.setOnClickListener(view -> {
                if (!astronaut.getInstagram().isEmpty()) {
                    startCustomTab(astronaut.getInstagram());
                } else {
                    Toast.makeText(mActivity, astronaut.getName().split(" ")[0] + " " + mActivity.getString(R.string.errorNoInstagram), Toast.LENGTH_SHORT).show();
                }
            });

            // Does the astronaut have a facebook handle?
            if (!astronaut.getFacebook().isEmpty()) {
                holder.mAstronautFacebook.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_facebook));
            } else {
                holder.mAstronautFacebook.setImageDrawable(ContextCompat.getDrawable(mActivity, R.drawable.ic_action_facebook_grey));
            }

            holder.mAstronautFacebook.setOnClickListener(view -> {
                if (!astronaut.getFacebook().isEmpty()) {
                    startCustomTab(astronaut.getFacebook());
                } else {
                    Toast.makeText(mActivity, astronaut.getName().split(" ")[0] + " " + mActivity.getString(R.string.errorNoFacebook), Toast.LENGTH_SHORT).show();
                }
            });

            // Does the astronaut have a wiki handle?
            holder.mAstronautWiki.setOnClickListener(view -> {
                if (astronaut.getWiki().length() != 0) {
                    startCustomTab(astronaut.getWiki());
                } else {
                    Toast.makeText(mActivity, astronaut.getName().split(" ")[0] + " " + mActivity.getString(R.string.errorNoWiki), Toast.LENGTH_SHORT).show();
                }
            });

            holder.mAstronautGoogle.setOnClickListener(view ->
                    startCustomTab("https://www.google.com/search?q=" + astronaut.getName()));

            /* Set the astronaut info and images */
            holder.mName.setText(astronaut.getName());
            holder.mRole.setText(astronaut.getRole() + " " + mActivity.getString(R.string.midAt) + " " + location);
            holder.mBio.setText(astronaut.getBio());

            try {
                holder.mDate.setText(mActivity.getString(R.string.since) + mDateFormat.format(new Date(astronaut.getLaunchDate() * 1000L)));
            } catch (Exception e) {
                holder.mDate.setVisibility(View.INVISIBLE);
            }

            Glide.with(mActivity).load(astronaut.getImage()).listener(new RequestListener<Drawable>() {
                @Override
                public boolean onLoadFailed(@Nullable GlideException e, @Nullable Object model, @NonNull Target<Drawable> target, boolean isFirstResource) {
                    holder.mAstronautPictureProgress.setVisibility(View.GONE);
                    return false;
                }

                @Override
                public boolean onResourceReady(@NonNull Drawable resource, @NonNull Object model, Target<Drawable> target, @NonNull DataSource dataSource, boolean isFirstResource) {
                    holder.mAstronautPictureProgress.setVisibility(View.GONE);
                    return false;
                }
            }).centerCrop().error(R.drawable.ic_failure_profile).into(holder.mProfileImage);

            Glide.with(mActivity).load("https://flagsapi.com/" + astronaut.getFlagCode() + "/flat/64.png").into(holder.mCountryFlag);
        }
    }

    public void startCustomTab(String url) {
        CustomTabsIntent intent = new CustomTabsIntent.Builder().setDefaultColorSchemeParams(
                new CustomTabColorSchemeParams.Builder()
                .setToolbarColor(ContextCompat.getColor(mActivity, R.color.colorPrimary))
                .build())
                .setUrlBarHidingEnabled(true)
                .setShowTitle(true)
                .build();
        intent.launchUrl(mActivity, Uri.parse(url));
    }

    /**
     * Size of our data base. Equals number of astronauts in space.
     *
     * @return An int representing the size
     */
    @Override
    public int getItemCount() {
        if (mAstronauts == null) return 0;
        return mAstronauts.size();
    }

    /**
     * Set the data list of recyclerview to a new one
     *
     * @param dataSet the new data list
     */
    @SuppressLint("NotifyDataSetChanged")
    public void setDataSet(List<Astronaut> dataSet) {
        mAstronauts = dataSet;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/restart/spacestationtracker/adapter/PeopleInSpaceAdapter.kt
package com.restart.spacestationtracker.adapter

import android.annotation.SuppressLint
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast

import androidx.annotation.DrawableRes
import androidx.appcompat.app.AppCompatActivity
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat

import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import com.restart.spacestationtracker.R
import com.restart.spacestationtracker.data.Astronaut
import com.restart.spacestationtracker.util.TopCrop

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

import androidx.recyclerview.widget.RecyclerView
import androidx.core.net.toUri

class PeopleInSpaceAdapter(
    private val activity: AppCompatActivity,
    private var listItems: List<Any>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val dateFormat = SimpleDateFormat("dd-MM-yyyy", Locale.getDefault())

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE_AD) {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.ad_native_layout, parent, false)
            AdViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.people_row, parent, false)
            PeopleInSpaceAdapterViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            VIEW_TYPE_ITEM -> {
                val itemHolder = holder as PeopleInSpaceAdapterViewHolder
                itemHolder.bind(listItems[position] as Astronaut)
            }
            VIEW_TYPE_AD -> {
                val adHolder = holder as AdViewHolder
                adHolder.bind(listItems[position] as NativeAd)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (listItems[position] is Astronaut) VIEW_TYPE_ITEM else VIEW_TYPE_AD
    }

    override fun getItemCount(): Int = listItems.size

    fun findPositionOf(astronaut: Astronaut): Int {
        return listItems.indexOf(astronaut)
    }

    private fun startCustomTab(url: String) {
        val intent = CustomTabsIntent.Builder().apply {
            setDefaultColorSchemeParams(
                CustomTabColorSchemeParams.Builder()
                    .setToolbarColor(ContextCompat.getColor(activity, R.color.colorPrimary))
                    .build()
            )
            setUrlBarHidingEnabled(true)
            setShowTitle(true)
        }.build()
        intent.launchUrl(activity, url.toUri())
    }

    inner class AdViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val adView: NativeAdView = view.findViewById(R.id.native_ad)
        private val adHeadline: TextView = view.findViewById(R.id.ad_headline)
        private val adAdvertiser: TextView = view.findViewById(R.id.ad_advertiser)
        private val adBody: TextView = view.findViewById(R.id.ad_body)
        private val adCallToAction: Button = view.findViewById(R.id.ad_call_to_action)
        private val adAppIcon: ImageView = view.findViewById(R.id.ad_app_icon)
        private val adMedia: MediaView = view.findViewById(R.id.ad_media)

        init {
            adView.headlineView = adHeadline
            adView.bodyView = adBody
            adView.callToActionView = adCallToAction
            adView.iconView = adAppIcon
            adView.advertiserView = adAdvertiser
            adView.mediaView = adMedia
        }

        fun bind(nativeAd: NativeAd) {
            adHeadline.text = nativeAd.headline
            adBody.text = nativeAd.body
            adCallToAction.text = nativeAd.callToAction
            adAdvertiser.text = nativeAd.advertiser

            nativeAd.icon?.let {
                adAppIcon.setImageDrawable(it.drawable)
                adAppIcon.visibility = View.VISIBLE
            } ?: run {
                adAppIcon.visibility = View.GONE
            }

            adView.setNativeAd(nativeAd)
        }
    }

    inner class PeopleInSpaceAdapterViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val profileImage: ImageView = view.findViewById(R.id.img)
        private val astronautPictureProgress: ProgressBar =
            view.findViewById(R.id.astronaut_picture_progress)
        private val countryFlag: ImageView = view.findViewById(R.id.countryFlag)
        private val astronautTwitter: ImageView = view.findViewById(R.id.astronautTwitter)
        private val astronautFacebook: ImageView = view.findViewById(R.id.astronautFacebook)
        private val astronautInstagram: ImageView = view.findViewById(R.id.astronautInstagram)
        private val astronautWiki: ImageView = view.findViewById(R.id.astronautWiki)
        private val astronautGoogle: ImageView = view.findViewById(R.id.astronautGoogle)
        private val name: TextView = view.findViewById(R.id.name)
        private val role: TextView = view.findViewById(R.id.role)
        private val date: TextView = view.findViewById(R.id.days_since_launch)
        private val bio: TextView = view.findViewById(R.id.bio)

        @SuppressLint("SetTextI18n")
        fun bind(astronaut: Astronaut) {
            val location = if (astronaut.isIss) "ISS" else "Tiangong"
            val firstName = astronaut.name.substringBefore(" ")

            setupSocialLink(
                astronautTwitter,
                astronaut.twitter,
                R.drawable.ic_action_twitter,
                R.drawable.ic_action_twitter_grey,
                "$firstName ${activity.getString(R.string.errorNoTwitter)}"
            )
            setupSocialLink(
                astronautInstagram,
                astronaut.instagram,
                R.drawable.ic_action_instagram,
                R.drawable.ic_action_instagram_grey,
                "$firstName ${activity.getString(R.string.errorNoInstagram)}"
            )
            setupSocialLink(
                astronautFacebook,
                astronaut.facebook,
                R.drawable.ic_action_facebook,
                R.drawable.ic_action_facebook_grey,
                "$firstName ${activity.getString(R.string.errorNoFacebook)}"
            )
            setupSocialLink(
                astronautWiki,
                astronaut.wiki,
                R.drawable.ic_wikipedia,
                R.drawable.ic_wikipedia_gray,
                "$firstName ${activity.getString(R.string.errorNoWiki)}"
            )

            astronautGoogle.setOnClickListener { startCustomTab("https://www.google.com/search?q=${astronaut.name}") }

            name.text = astronaut.name
            role.text = "${astronaut.role} ${activity.getString(R.string.midAt)} $location"
            bio.text = astronaut.bio

            try {
                date.text =
                    activity.getString(R.string.since) + dateFormat.format(Date(astronaut.launchDate * 1000L))
                date.visibility = View.VISIBLE
            } catch (e: Exception) {
                date.visibility = View.INVISIBLE
            }

            Glide.with(activity).load(astronaut.image).listener(object : RequestListener<Drawable> {
                override fun onLoadFailed(
                    e: GlideException?,
                    model: Any?,
                    target: Target<Drawable>,
                    isFirstResource: Boolean
                ): Boolean {
                    astronautPictureProgress.visibility = View.GONE
                    return false
                }

                override fun onResourceReady(
                    resource: Drawable,
                    model: Any,
                    target: Target<Drawable>,
                    dataSource: DataSource,
                    isFirstResource: Boolean
                ): Boolean {
                    astronautPictureProgress.visibility = View.GONE
                    return false
                }
            }).transform(TopCrop()).error(R.drawable.ic_failure_profile).into(profileImage)

            Glide.with(activity).load("https://flagsapi.com/${astronaut.flagCode}/flat/64.png")
                .into(countryFlag)
        }

        private fun setupSocialLink(
            imageView: ImageView,
            url: String,
            @DrawableRes activeIconRes: Int,
            @DrawableRes inactiveIconRes: Int,
            errorToastString: String
        ) {
            val isActive = url.isNotEmpty()
            imageView.setImageResource(if (isActive) activeIconRes else inactiveIconRes)
            imageView.setOnClickListener {
                if (isActive) {
                    startCustomTab(url)
                } else {
                    Toast.makeText(activity, errorToastString, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    companion object {
        private const val VIEW_TYPE_ITEM = 0
        private const val VIEW_TYPE_AD = 1
    }
}